### PR TITLE
Add option to set the ellipsoid to use for calculations with qgis_process tool

### LINKED ISSUE #37988: qgis_process: impossible to execute some algorithms for lack of project context
**Describe the bug**

Some algorithms need information stored in the project properties, like ellipsoid to use, or crs of the project, area and length measurement units, in order to properly execute the job.

It seems impossible to run those algs using qgis_process because it is not possible to specify a project containing the needed information nor to provide the necessary information directly passing them as additional parameters.

A probably not comprehensive list of algorithms that use context.project() follows:

qgis:exportaddgeometrycolumns (only methods 1 e 2)
qgis:fieldcalculator
qgis:distancetonearesthublinetohub
qgis:distancetonearesthubpoints
qgis:distancematrix
qgis:pointstopath
qgis:randompointsalongline
qgis:randompointsinsidepolygons
qgis:executesql
qgis:rastercalculator
qgis:postgisexecuteandloadsql
qgis:tilesxyzdirectory

native:antimeridiansplit
native:sumlinelengths
native:calculatevectoroverlaps
native:nearestneighbouranalysis
native:refactorfields
native:linedensity
native:aggregate
native:hublines
native:categorizeusingstyle
native:networkanalysis
native:rasterize
native:loadlayer
native:setprojectvariable
native:stylefromproject
native:bookmarkstolayer

**How to Reproduce**

```
qgis_process run qgis:randompointsinsidepolygons --STRATEGY=0 --VALUE=1000 --INPUT="./ProvCM01012020_g/ProvCM01012020_g_WGS84.shp" --OUTPUT="./output/random_point_prov.shp"

/usr/lib/python3/dist-packages/qgis/utils.py:743: DeprecationWarning: the
imp module is deprecated in favour of importlib; see the module's
documentation for alternative uses
  mod = _builtin_import(name, globals, locals, fromlist, level)
Problem with SAGA installation: SAGA was not found or is not correctly
installed
----------------
Inputs
----------------

INPUT:  ./ProvCM01012020_g/ProvCM01012020_g_WGS84.shp
OUTPUT: ./output/random_point_prov.shp
STRATEGY:       0
VALUE:  1000

Logged critical: Traceback (most recent call last):
  File
"/usr/share/qgis/python/plugins/processing/algs/qgis/RandomPointsPolygons.py",
line 160, in processAlgorithm
    da.setEllipsoid(context.project().ellipsoid())
AttributeError: 'NoneType' object has no attribute 'ellipsoid'

Traceback (most recent call last):
  File
"/usr/share/qgis/python/plugins/processing/algs/qgis/RandomPointsPolygons.py",
line 160, in processAlgorithm
    da.setEllipsoid(context.project().ellipsoid())
AttributeError: 'NoneType' object has no attribute 'ellipsoid'
ERROR:  Traceback (most recent call last):
  File
"/usr/share/qgis/python/plugins/processing/algs/qgis/RandomPointsPolygons.py",
line 160, in processAlgorithm
    da.setEllipsoid(context.project().ellipsoid())
AttributeError: 'NoneType' object has no attribute 'ellipsoid'
```

**QGIS and OS versions**

QGIS 3.14.1

**Additional context**

Ref: https://github.com/qgis/QGIS/pull/34617

### PR DESCRIPTION
**Pull request by nyalldawson**
- Add explicit settings for ellipsoid/distance unit/area unit to QgsProcessingContext, and prefer to use these instead of trying to get the settings from a QgsProject (which isn't available from the standalone tool).
- Expose a new argument to allow users to specify "--ELLIPSOID=xxxx" when running algorithms via the qgis_process tool.
- Flags algorithms which make no sense to run outside of the GUI environment (such as "select by ..." algorithms), and hides/prevents them from running via qgis_process.
- Also warns when running deprecated or algorithms with known issues from qgis_process
- Flags algorithms which require a valid project to execute so that errors are raised when running these from qgis_process, unless the new "--PROJECT_PATH=xxxx" argument is specified and pointing to a valid project path.

Fixes #37988

### CODE REVIEW

File: python/plugins/processing/algs/qgis/ExportGeometryInfo.py [hunk 1]
```diff
@@ -138,7 +138,7 @@ def processAlgorithm(self, parameters, context, feedback):
         self.distance_area = QgsDistanceArea()
         if method == 2:
             self.distance_area.setSourceCrs(source.sourceCrs(), context.transformContext())
-            self.distance_area.setEllipsoid(context.project().ellipsoid())
+            self.distance_area.setEllipsoid(context.ellipsoid())
         elif method == 1:
             coordTransform = QgsCoordinateTransform(source.sourceCrs(), context.project().crs(), context.project())
```
Comment (agiudiceandrea) [inline python/plugins/processing/algs/qgis/ExportGeometryInfo.py:143]: Hi @nyalldawson, this algorithm also needs `context.project().crs()` and `context.project()`

`context.project().crs()` is also needed for example by [native:networkanalysis](https://github.com/qgis/QGIS/blob/master/src/analysis/processing/qgsalgorithmnetworkanalysisbase.cpp#L116)

What about some other algs that use the whole context.project() like for example native:rasterize native:loadlayer, native:setprojectvariable, native:stylefromproject, native:bookmarkstolayer, native:categorizeusingstyle, ... ?

Comment (nyalldawson) [reply python/plugins/processing/algs/qgis/ExportGeometryInfo.py:143]: Yes, it's only a partial fix but addresses the immediate ellipsoid handling need. 

I guess we also need a --project=... argument , and then make sure this is specified if an algorithm requires it. That said, "export geometry info" is a generally badly behaved algorithm and needs a rewrite (in C++). And some of those other algorithms have no real meaning outside of qgis and should be hidden entirely (such as the "select by...." algorithms). I'll add a flag to hide algorithms from the standalone tool to handle this.

Comment (agiudiceandrea) [reply python/plugins/processing/algs/qgis/ExportGeometryInfo.py:143]: Thank you very much!

Comment (nyalldawson) [reply python/plugins/processing/algs/qgis/ExportGeometryInfo.py:144]: Ok, I've updated this PR with the extra logic now...

**Comment by nyalldawson** [comment]
Unrelated test failure

**Comment by Gustry** [comment]
Nice addition !
Can you add the new project option in the PR description (for changelog and doc)? I can't edit. Thanks

**Comment by nyalldawson** [comment]
@Gustry done

**Comment by agiudiceandrea** [comment]
@nyalldawson did you forget to mention the "**--AREA_UNIT**=" and "**--DISTANCE_UNIT**=" parameters in the PR description and in the usage description? Or are they not intended to be used?

**Pull request merged**